apps/src/lib/util/firehose.js

```diff
@@ -311,35 +312,65 @@ class FirehoseClient {
 // This code sets up an AWS config against a very restricted user, so this is
 // not a concern, we just don't want to make things super obvious. For the
 // plaintext, contact asher or eric.
-// eslint-disable-next-line
-const _0x12ed = [
-  '\x41\x4b\x49\x41\x4a\x41\x41\x4d\x42\x59\x4d\x36\x55\x53\x59\x54\x34\x35\x34\x51',
-  '\x78\x4e\x4e\x39\x4e\x79\x32\x61\x6d\x39\x78\x75\x4b\x79\x57\x39\x53\x2b\x4e\x76\x41\x77\x33\x67\x68\x68\x74\x68\x72\x6b\x37\x6b\x6e\x51\x59\x54\x77\x6d\x4d\x48',
-  '\x75\x73\x2d\x65\x61\x73\x74\x2d\x31',
-  '\x63\x6f\x6e\x66\x69\x67'
-];
-(function(_0xb54a92, _0x4e682a) {
-  var _0x44f3e8 = function(_0x35c55a) {
-    while (--_0x35c55a) {
-      _0xb54a92['\x70\x75\x73\x68'](_0xb54a92['\x73\x68\x69\x66\x74']());
-    }
+function createNewFirehose(AWS, Firehose) {
+  // eslint-disable-next-line
+  const _0x12ed = [
+    '\x41\x4b\x49\x41\x4a\x41\x41\x4d\x42\x59\x4d\x36\x55\x53\x59\x54\x34\x35\x34\x51',
+    '\x78\x4e\x4e\x39\x4e\x79\x32\x61\x6d\x39\x78\x75\x4b\x79\x57\x39\x53\x2b\x4e\x76\x41\x77\x33\x67\x68\x68\x74\x68\x72\x6b\x37\x6b\x6e\x51\x59\x54\x77\x6d\x4d\x48',
+    '\x75\x73\x2d\x65\x61\x73\x74\x2d\x31',
+    '\x63\x6f\x6e\x66\x69\x67'
+  ];
+  (function(_0xb54a92, _0x4e682a) {
+    var _0x44f3e8 = function(_0x35c55a) {
+      while (--_0x35c55a) {
+        _0xb54a92['\x70\x75\x73\x68'](_0xb54a92['\x73\x68\x69\x66\x74']());
+      }
+    };
+    _0x44f3e8(++_0x4e682a);
+  })(_0x12ed, 0x127);
+  var _0xd12e = function(_0x2cedd5, _0x518781) {
+    _0x2cedd5 = _0x2cedd5 - 0x0;
+    var _0x4291ea = _0x12ed[_0x2cedd5];
+    return _0x4291ea;
   };
-  _0x44f3e8(++_0x4e682a);
-})(_0x12ed, 0x127);
-var _0xd12e = function(_0x2cedd5, _0x518781) {
-  _0x2cedd5 = _0x2cedd5 - 0x0;
-  var _0x4291ea = _0x12ed[_0x2cedd5];
-  return _0x4291ea;
-};
-AWS[_0xd12e('0x0')] = new AWS['\x43\x6f\x6e\x66\x69\x67']({
-  accessKeyId: _0xd12e('0x1'),
-  secretAccessKey: _0xd12e('0x2'),
-  region: _0xd12e('0x3')
-});
+  AWS[_0xd12e('0x0')] = new AWS['\x43\x6f\x6e\x66\x69\x67']({
+    accessKeyId: _0xd12e('0x1'),
+    secretAccessKey: _0xd12e('0x2'),
+    region: _0xd12e('0x3')
+  });
+
+  return new Firehose({
+    apiVersion: '2015-08-04',
+    correctClockSkew: true
+  });
+}
+
+let promise;
+function getSingleton() {
+  return Promise.all([
+    import('aws-sdk/lib/core'),
+    import('aws-sdk/clients/firehose'),
+    import('aws-sdk/lib/config')
+  ]).then(([{default: AWS}, {default: Firehose}]) => {
+    if (!promise) {
+      promise = new Promise(resolve =>
+        resolve(new FirehoseClient(AWS, Firehose))
```

Comment:
won't this reimport every time `getSingleton` is called? I would expect this to look like

```js
let firehoseClient;
function getSingleton() {
  if (firehoseClient) {
    return new Promise(resolve => resolve(firehoseClient));
  }

  return Promise.all([
    import("aws-sdk/lib/core"),
    import("aws-sdk/clients/firehose"),
    import("aws-sdk/lib/config")
  ]).then(([{ default: AWS }, { default: Firehose }]) => {
    firehoseClient = new FirehoseClient(AWS, Firehose);
    return firehoseClient;
  });
}
```

Reply:
Ah, good point! I agree about not importing multiple times, however I'm worried that the specific alternative you've provided might lead to two `FirehoseClient` objects being initialized if a second request for `getSingleton` arrived before the first was complete. I've pushed an update which I think gives us a guarantee that we only import once and also only create one `FirehoseClient`. 

I haven't explicitly stepped through to verify this, but I've at least verified that when sending multiple sequential firehose events, the files are only fetched once from the network, and multiple log messages are generated.